locator/topology.hh

```diff
@@ -114,15 +217,30 @@ public:
     void sort_by_proximity(inet_address address, inet_address_vector_replica_set& addresses) const;
 
 private:
+    using mutable_node_ptr = lw_shared_ptr<node>;
+
     // default constructor for cloning purposes
-    topology() = default;
+    topology() noexcept;
+
+    node_ptr add_node(mutable_node_ptr node);
+    void remove_node(node_ptr node);
+    void do_remove_node(mutable_node_ptr node);
 
     /**
      * compares two endpoints in relation to the target endpoint, returning as
      * Comparator.compare would
      */
     int compare_endpoints(const inet_address& address, const inet_address& a1, const inet_address& a2) const;
 
+    unsigned _shard;
+    std::unordered_set<mutable_node_ptr> _all_nodes;
+    node_ptr _local_node;
+    std::unordered_map<host_id, const node*> _nodes_by_host_id;
+    std::unordered_map<inet_address, const node*> _nodes_by_endpoint;
+
+    std::unordered_map<sstring, std::unordered_set<const node*>> _dc_nodes;
```

Comment:
Probably, std::vector<> is a better fit here. Depends on how it's used.

Reply:
Why would a vector be better? for memory density?
Although unordered_set has cheaper delete when a node is removed, it is a rare event and maybe memory efficiency trumps cpu efficiency in this case.

Reply:
the `_all_nodes` map is changed to `std::vector<unique_ptr<node>> _nodes` in v8
by dc/rack map kept the same.

Eventually I'd like to send a follow up that defines `datacenter` and `rack` classes and let nodes point to respective objects rather than return dc_rack by name.
and then the maps will contain those datacenter and rack objects. that may contain a vector of nodes, although to modify them, it will still be faster keeping an `unordered_set<const node*>` rather than a `vector<const node*>`

Reply:
Yes memory density, and efficiency. We regenerate the whole thing when a node is dropped, no?

Reply:
Yes, we regenerate the whole thing when a node is dropped or added or replaced.